File: exercism/python/bob/bob.py
from string import whitespace, ascii_letters, ascii_uppercase


def is_question(phrase):
    return '?' in phrase and is_empty(phrase[len(phrase)-phrase[::-1].find('?'):])


def is_yell(phrase):
    return any(map(lambda c: c in ascii_letters, phrase)) and \
        all(map(lambda c: c not in ascii_letters or c in ascii_uppercase, phrase))


def is_empty(phrase):
    return all(map(lambda c: c in whitespace, phrase))


def hey(phrase):
    if is_question(phrase) and is_yell(phrase):
        return "Calm down, I know what I'm doing!"
    if is_question(phrase):
        return 'Sure.'
    if is_yell(phrase):
        return 'Whoa, chill out!'
    if is_empty(phrase):
        return 'Fine. Be that way!'
    return 'Whatever.'

File: exercism/elixir/bob/lib/bob.ex
# https://exercism.io/my/solutions/2966a9d0d2bf4f00bced521ddaebb15c

defmodule Bob do
  @spec is_question?(String.t()) :: boolean()
  def is_question?(s), do: String.ends_with?(s, "?")

  @spec is_yell?(String.t()) :: boolean()
  defp is_yell?(s), do: String.upcase(s) == s and String.downcase(s) != s

  @spec is_empty?(String.t()) :: boolean()
  defp is_empty?(s), do: s == ""

  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed = String.trim(input)

    cond do
      is_question?(trimmed) and is_yell?(trimmed) -> "Calm down, I know what I'm doing!"
      is_question?(trimmed) -> "Sure."
      is_yell?(trimmed) -> "Whoa, chill out!"
      is_empty?(trimmed) -> "Fine. Be that way!"
      true -> "Whatever."
    end
  end
end
